### ISSUE TITLE
Macedonian (mkd): unicharset omits U+0455 (dze) and U+0459 (lje) -- two letters of the alphabet cannot be output

### ISSUE THREAD

**Issue by inventif** [tesseract-ocr/tessdata#204]
# Macedonian (`mkd`) model cannot output two letters of the alphabet: `ѕ` (U+0455) and `љ` (U+0459)

**Model:** `mkd.traineddata`, version string `best2int20180322`
**Affects:** `tessdata`, `tessdata_best`, `tessdata_fast` — all three
**Tesseract tested:** 5.5.3.20260724 (Windows), also reproduced against `tessdata_best` and `tessdata_fast` models
**Severity:** two letters of a national alphabet are unrepresentable; the corruption is silent

---

## Summary

The Macedonian model's unicharset omits **dze** (`ѕ` U+0455 / `Ѕ` U+0405) and **lje**
(`љ` U+0459 / `Љ` U+0409). Because the LSTM has no output class for them, it **cannot
emit them at any confidence**. This is not a recognition-accuracy issue that better
images or different `--psm` settings can improve.

Both are ordinary letters of the modern Macedonian alphabet:

- `љ` is common — *љубов* (love), *љубезен* (kind), *Љубљана* (Ljubljana)
- `ѕ` is rarer but irreplaceable — *ѕид* (wall), *ѕвезда* (star), *ѕвер* (beast), *соѕвездие* (constellation)

**The `љ` failure is silent and therefore the more damaging of the two:** `љ` is
consistently emitted as `њ`, which is *also* a real Macedonian letter. The output stays
plausible, passes a spellchecker's eye, and a reader who does not have the source text
cannot tell the text has been corrupted.

### 1. The characters are absent from the unicharset

```
combine_tessdata -u mkd.traineddata mkd.
```

| Component | entries | contains `ѕ` U+0455 | contains `љ` U+0459 |
|---|---|---|---|
| `unicharset` (legacy) | 95 | **no** | yes |
| `lstm-unicharset` | 85 | **no** | **no** |

The `lstm-unicharset` contains 58 Cyrillic characters, including the other
Macedonian-specific letters `ѓ` `ј` `њ` `ќ` `џ` — but not `ѕ` or `љ`.

### 2. The unicharset predicts real behaviour exactly (8/8)

Each Macedonian-specific letter was rendered in a real word and OCR'd with `-l mkd`.
The prediction from reading the unicharset matched observed behaviour for **all eight**:

| Letter | in `lstm-unicharset` | OCR result |
|---|---|---|
| `ѓ` U+0453 | yes | correct |
| `ј` U+0458 | yes | correct |
| `њ` U+045A | yes | correct |
| `ќ` U+045C | yes | correct |
| `џ` U+045F | yes | correct |
| `з` U+0437 (control) | yes | correct |
| **`ѕ` U+0455** | **no** | **fails** |
| **`љ` U+0459** | **no** | **fails** |

### 3. Not font-dependent and not flavour-dependent

`ѕвезда` was rendered in 8 fonts, each verified via the font's `cmap` to actually
contain U+0455 and U+0409 (so `.notdef` rendering is excluded as a cause):

**0 of 8 fonts** recovered the `ѕ`. What it degrades into varies by font, which is
itself evidence of an unlearned glyph rather than a systematic substitution:

| Output | Fonts |
|---|---|
| `5везда` | Calibri, Consolas, Segoe UI, Verdana |
| `звезда` | Courier New, Georgia, Times New Roman |
| `ввезда` | Arial (`--psm 7`) |

**0 of 3 tessdata flavours** recovered it either — `tessdata`, `tessdata_best` and
`tessdata_fast` all fail, while the control word `звезда` (ordinary `з`) is read
correctly by all three.

`љубов` is read as `њубов` in every font tested.

### 4. The shipped dictionaries are corrupted in the same direction

Extracted with `dawg2wordlist`:

| Component | words | with `ѕ` | with `љ` | with `њ` |
|---|---|---|---|---|
| `word-dawg` (legacy) | 116,522 | **0** | 247 | 4,298 |
| `lstm-word-dawg` | 108,883 | **0** | **0** | 4,189 |

Zero `љ` words against 4,189 `њ` words in the LSTM dictionary. Macedonian does not look
like this. For comparison, a corpus built from the Macedonian Wikipedia article
namespace (222,971 articles, 2.89M sentences) plus the hunspell `mk` dictionary yields
**2,642 distinct words containing `ѕ`** and **7,121 containing `љ`**, at a `љ`/`њ` ratio
of 0.23 — against the shipped model's 0.00.

This suggests the training corpus collapsed `љ` into `њ` and dropped `ѕ` entirely, and
that the defect is in the source data rather than in training.

## Reproduction

```bash
# 1. static: the characters are not in the charset
combine_tessdata -u mkd.traineddata mkd.
grep -c $'ѕ' mkd.lstm-unicharset   # 0
grep -c $'љ' mkd.lstm-unicharset   # 0

# 2. behavioural: render "ѕвезда" (star) in any font containing U+0455
tesseract zvezda.png stdout -l mkd --psm 7
#   expected: ѕвезда
#   actual:   5везда / звезда / ввезда  (varies by font)

# 3. the silent one: render "љубов" (love)
tesseract ljubov.png stdout -l mkd --psm 7
#   expected: љубов
#   actual:   њубов      <-- a real Macedonian letter, so the error is invisible
```

## Suggested fix

Rebuild `mkd` with a corpus and unicharset covering the full Macedonian alphabet, and
regenerate the word DAWGs from a wordlist that contains `ѕ` and `љ` words. Fixing the
unicharset alone is not sufficient — the shipped dictionaries would still contain zero
instances of either letter and would work against the restored characters at inference.

## Reporter

Marijančo Galevski (Маријанчо Галевски)

Scripts reproducing every measurement above are available on request; they depend only
on Tesseract's own tooling (`combine_tessdata`, `dawg2wordlist`) plus PIL for rendering.

**Comment by inventif** [comment]
Follow-up: I retrained the model to test whether the diagnosis above actually holds, and it does. Publishing the result here in case it is useful as corroboration or as a starting point — not as a request to adopt these weights.

**Corrected model:** https://github.com/inventif/ljubomir (Apache-2.0, [v1.0.0](https://github.com/inventif/ljubomir/releases/tag/v1.0.0))

## What this confirms

Fine-tuning from `tessdata_best` with the unicharset extended by exactly four characters (`ѕ Ѕ љ Љ`, 86 → 90 classes) and the word DAWGs rebuilt from a corpus containing both letters restores them. That is consistent with the original diagnosis: the letters were missing from the charset, not merely hard to recognise.

Measured on held-out renders, both models through an identical code path, single evaluation run:

| | stock | retrained |
|---|---|---|
| `ѕ` U+0455 presence, ordinary typefaces | **0.0%** | 99.4% |
| `љ` U+0459 presence, ordinary typefaces | **0.0%** | 97.5% |
| exact word reads (27 words × 10 faces × 4 conditions × 2 font groups) | 1026/2160 | 1860/2160 |
| line CER, ordinary typefaces (1,275 lines) | 0.0336 | 0.0291 |

The sharpest case is three words that need the restored letter **twice in a single word** — `ѕрѕало` (mirror), `ѕрнѕало`, `ѕенѕало` — none of which appears in any of the 222,971 Wikipedia articles the corpus was built from. Stock **0/48**, retrained **48/48**.

`ѕрѕало` is the strongest of the three, because it is also absent from the retrained model's **own dictionary** — it was contributed after the wordlist was built and deliberately left out. It therefore has neither corpus support nor lexical support, so a correct reading of it cannot be memorisation or dictionary lookup; it is letterform recognition alone. On that word: stock **0/16**, retrained **16/16**. Stock cannot produce the letter even once, returning `врвало` / `зрвало`.

## What it costs, stated plainly

1. **`њ` U+045A regresses**, 100% → 96.0%. This looks intrinsic rather than fixable by more data: doubling the `љ`/`њ` contrast set did not remove it, it only moved along the trade-off (one checkpoint recovered `њ` to 98.5% while `ѕ` fell to 92.6%).
2. **CJK typefaces carrying Cyrillic get worse** — line CER 0.0856 → 0.1890 on Yu Gothic faces, from spurious spaces rather than wrong letters. Ordinary typefaces are unaffected and slightly better.
3. My strict release gate therefore returns **NO-GO** on (1), and I have published that verdict rather than relaxing the gate.
4. Evaluation is synthetic — rendered fonts, not scanned pages.

## The part that seems most relevant upstream

The fix that matters is probably not the weights but the **training data**. The shipped `lstm-word-dawg` contains 4,189 words with `њ` and **zero** with `љ`, which is not what Macedonian looks like. A corpus built from the Macedonian Wikipedia article namespace plus the hunspell `mk` dictionary yields **7,121 distinct words containing `љ`** and **2,642 containing `ѕ`**, a `љ`/`њ` ratio of 0.23 against the shipped model's 0.00.

So extending the unicharset alone would not be enough: the language model would still carry zero instances of either letter and would work against the restored characters at inference. Both the corpus builder and the evaluation harness are in the repo above if any of it is useful.

---

*The model is dedicated to my father, Ljubomir — Љубомир — who died on 1996-08-08, thirty years to the day before this work began. His name begins with `Љ`, the letter the stock model cannot write. The dedication is carried inside the model itself, in its `version` component.*
